File: race_officer/lib/features/race/domain/race_competitor.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:sailbrowser_flutter/features/race/domain/result_code.dart';

part 'race_competitor.freezed.dart';
part 'race_competitor.g.dart';

///
@freezed
class RaceCompetitor with _$RaceCompetitor {
  factory RaceCompetitor({
    required String id,
    required String raceId,
    required String seriesId,
    required String helm,
    String? crew,
    required String boatClass,
    required int sailNumber,
    required double handicap,

    /// Finish time recored when competitor finishes.
    /// If a manual finish time set by hand is specified it is used in preference.
    /// The recorded finish time is null if it has not been recorded.
    DateTime? recordedFinishTime,

    /// Manually specified finsih time.
    /// Used in preference to the recorded finish time if specified.
    /// The manual finish time may be set to null to disable it.
    DateTime? manualFinishTime,

    /// Start time for the competitor - set based on the race at start and finish
    DateTime? startTime,
    @Default(ResultCode.notFinished) ResultCode resultCode,

    /// Number of laps - defaults to number of lap times but may be manually set
    @Default(0) int manualLaps,
    @Default([]) List<DateTime> lapTimes,
    ResultData? result,
  }) = _RaceCompetitor;

  const RaceCompetitor._();

  factory RaceCompetitor.fromJson(Map<String, Object?> json) =>
      _$RaceCompetitorFromJson(json);

  /// The number of laps, manual value if set, otherwise the number of lap times recorded.
  int get numLaps {
    if (manualLaps != 0) {
      return manualLaps;
    } else {
      // If competitor has finished then he has completed an extra lap.
      return (finishTime == null) ? lapTimes.length : lapTimes.length + 1;
    }
  }

  /// Gets the finish time using a manually entered time in preference to 
  /// a recorded one.  Returns null if no finish time is avalaible.
  DateTime? get finishTime =>
      (manualFinishTime != null) ? manualFinishTime : recordedFinishTime;

  /// Returns the total time taken (finish - start).   
  /// This is not the same as the elapsed time that is 
  Duration get totalTime => (finishTime != null && startTime != null)
      ? finishTime!.difference(startTime!)
      : const Duration();

  String get helmCrew =>
      (crew != null && crew!.trim().isNotEmpty) ? '$helm / $crew}' : helm;

  /// Competitor has finished OK
  get isOk => resultCode == ResultCode.ok;
}

/// Results data that depends on other competitors in the race.
@freezed
class ResultData with _$ResultData {
  factory ResultData({
    @Default(Duration()) Duration elapsedTime,
    @Default(Duration()) Duration correctedTime,
    required int position,
    @Default(0) double points,
    required bool isDiscarded,
    required bool isDiscardable,
  }) = _ResultData;

  const ResultData._();

  factory ResultData.fromJson(Map<String, Object?> json) =>
      _$ResultDataFromJson(json);
}

File: sailbrowser-web/src/app/race/@store/race-competitor.ts
import { ResultCode } from './result-code';

export interface ResultData {
  elapsedTime: number; // milliseconds
  correctedTime: number; // milliseconds
  position: number;
  points: number;
  isDiscarded: boolean;
  isDiscardable: boolean;
}

export class RaceCompetitor {
  id: string;
  raceId: string;
  seriesId: string;
  helm: string;
  crew?: string;
  boatClass: string;
  sailNumber: number;
  handicap: number;

  /**
   * Finish time recorded when competitor finishes.
   * If a manual finish time set by hand is specified it is used in preference.
   * The recorded finish time is null if it has not been recorded.
   */
  recordedFinishTime?: Date;

  /**
   * Manually specified finish time.
   * Used in preference to the recorded finish time if specified.
   * The manual finish time may be set to null to disable it.
   */
  manualFinishTime?: Date;

  /**
   * Start time for the competitor - set based on the race at start and finish
   */
  startTime?: Date;

  resultCode: ResultCode;

  /**
   * Number of laps - defaults to number of lap times but may be manually set
   */
  manualLaps: number;
  lapTimes: Date[];
  result?: ResultData;

  constructor(data: Partial<RaceCompetitor>) {
    this.id = data.id || '';
    this.raceId = data.raceId || '';
    this.seriesId = data.seriesId || '';
    this.helm = data.helm || '';
    this.crew = data.crew;
    this.boatClass = data.boatClass || '';
    this.sailNumber = data.sailNumber || 0;
    this.handicap = data.handicap || 0;
    this.recordedFinishTime = data.recordedFinishTime;
    this.manualFinishTime = data.manualFinishTime;
    this.startTime = data.startTime;
    this.resultCode = data.resultCode || ResultCode.NotFinished;
    this.manualLaps = data.manualLaps || 0;
    this.lapTimes = data.lapTimes || [];
    this.result = data.result;
  }
}

import { Pipe, PipeTransform } from "@angular/core";

@Pipe({
  name: 'helmcrew',
})
export class HelmCrew implements PipeTransform {
  transform(comp: RaceCompetitor | null) {
    if (!comp) return;
    return comp.crew && comp.crew.trim().length > 0 ? `${comp.helm} / ${comp.crew}` : comp.helm;
  }
} 

  /**
   * TODO functions originall part of the competitor - convert to utility functions 
   * 
  /**
   * The number of laps, manual value if set, otherwise the number of lap times recorded.
   
  get numLaps(): number {
  if (this.manualLaps !== 0) {
    return this.manualLaps;
  } else {
    // If competitor has finished then he has completed an extra lap.
    return this.finishTime === undefined ? this.lapTimes.length : this.lapTimes.length + 1;
  }
   * 
   * Gets the finish time using a manually entered time in preference to
   * a recorded one. Returns undefined if no finish time is available.
   
  get finishTime(): Date | undefined {
  return this.manualFinishTime !== undefined && this.manualFinishTime !== null
    ? this.manualFinishTime
    : this.recordedFinishTime;
}

  /**
   * Returns the total time taken (finish - start) in milliseconds.
   
  get totalTime(): number {
  if (this.finishTime && this.startTime) {
    return this.finishTime.getTime() - this.startTime.getTime();
  }
  return 0;
}

  get helmCrew(): string {
  return this.crew && this.crew.trim().length > 0 ? `${this.helm} / ${this.crew}` : this.helm;
}

  /**
   * Competitor has finished OK
   
  get isOk(): boolean {
  return this.resultCode === ResultCode.Ok;
}
  */
